File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        max_window = 0
        if len(s) < 2:
            return len(s)
        for right in range(len(s)):
            if s[right] in s[left:right]:
                left += s[left:right].index(s[right]) + 1
            max_window = max(max_window, right - left + 1)
        return max_window

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap();
        int right = 0, max = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) {
                right = right> map.get(c)+1? right:map.get(c)+1;
            }
            map.put(c, i);
            max = max >  i - right +1 ? max :  i - right +1;
        }
        return max;
    }
}